docs/configuration.md
# Configuration

CLI options and environment vars are equal in their usage. If both a CLI flag and an environment variable is set, the CLI flag takes precedence. For boolean CLI flags, they have no value. For environment vars, set them to either `true` or `false`, other values will not be accepted, and omitting them will set them to `false`.

## Required

- `--key` / `CAMO_KEY` - Randomly generated string used as a key for calculating the HMAC digest.
- `--root-url` / `CAMO_ROOT_URL` - URL, including a trailing slash, relative to the domain Camo is running on. For example, if Camo is available on `example.com/camo/`, set this to `/camo/`. For installations that do not run in a subdirectory, set this to `/`.

## Allowed content-types

At least one `content-type` needs to be allowed, or Camo will refuse to start.

- `--allow-audio` / `CAMO_ALLOW_AUDIO` - Whether `audio/*` MIME types should be allowed. (default: `false`)
- `--allow-image` / `CAMO_ALLOW_IMAGE` - Whether `image/*` MIME types should be allowed. (default: `false`)
- `--allow-json` / `CAMO_ALLOW_JSON` - Whether `application/json*` MIME types should be allowed. (default: `false`)
- `--allow-video` / `CAMO_ALLOW_VIDEO` - Whether `video/*` MIME types should be allowed. (default: `false`)
- `--allow-all` / `CAMO_ALLOW_ALL` - Accepts any content type. (default: `false`)

## Other settings

- `--header-via` / `CAMO_HEADER_VIA` - The string used to identify this `camo-rs` instance in upstream requests. (default: `camo-rs asset proxy (+https://github.com/denschub/camo-rs)`)
- `--length-limit` / `CAMO_LENGTH_LIMIT` - The maximum `content-length` proxied by `camo-rs`. (default: `52428800` (50 MiB))
- `--listen` / `CAMO_LISTEN` - IP and Port this application should listen on. (default: `[::]:8081`)
- `--upstream-timeout` / `CAMO_UPSTREAM_TIMEOUT` - The number of seconds to wait for an upstream response. (default: `10`)

## Logging

By default, `camo-rs` is very quiet. It will only ever say anything if something goes wrong. Optional logging is available.

### Log levels

The `--log-level` flag or `CAMO_LOG_LEVEL` env var can have the following values:

- `quiet` - Doesn't log anything at all, unless something unexpected goes wrong. This is the default.
- `warn` - Logs when valid requests couldn't be processed due to upstream errors, or if requests have been blocked by length limits or content-type restrictions.
- `info` - Logs the same was `warn`, but additionally logs when the request field encoding was wrong, or if the HMAC was invalid.

### Log formats

The `--log-format` flag or `CAMO_LOG_FORMAT` env var can be set to:

- `text` - Does log the information in a human-readable format.
- `text-color` - The same as `text`, but with some color added to improve readability. This is the default.
- `json` - Logs the data in a machine-readable JSON format, with one log entry per line.


src/settings.rs
//! The Application Settings Module(tm)

use tracing::Level;

/// Specifies the log's output format
#[derive(clap::ValueEnum, Clone, Debug)]
pub enum LogFormat {
    /// Logs in a human-readable format.
    Text,

    /// Logs in a human-readable format, but with colors.
    TextColor,

    /// Logs into a machine-readable JSONL format.
    Json,
}

/// Specifies how much log output Camo generates
#[derive(clap::ValueEnum, Clone, Debug)]
pub enum LogLevel {
    /// Does not log anything at all
    Quiet,

    /// Logs upstream errors and blocked upstream requests
    Warn,

    /// Everything that `warn` includes, but also logs invalid URLs/HMACs
    Info,
}

impl LogLevel {
    pub fn tracing_level(&self) -> Level {
        match self {
            LogLevel::Quiet => Level::ERROR,
            LogLevel::Warn => Level::WARN,
            LogLevel::Info => Level::INFO,
        }
    }
}

/// Application Settings Struct, designed to be primarily used by Clap
#[derive(clap::Parser, Clone, Debug)]
#[clap(about, author, version = env!("CAMO_RS_VERSION"))]
pub struct Settings {

    /// If present, all content types will be allowed
    #[clap(long = "allow-all", env = "CAMO_ALLOW_ALL")]
    pub allow_all: bool,

    /// If present, `audio/*` MIME types will be allowed
    #[clap(long = "allow-audio", env = "CAMO_ALLOW_AUDIO")]
    pub allow_audio: bool,

    /// If present, `image/*` MIME types will be allowed
    #[clap(long = "allow-image", env = "CAMO_ALLOW_IMAGE")]
    pub allow_image: bool,

    /// If present, `json/*` MIME types will be allowed
    #[clap(long = "allow-json", env = "CAMO_ALLOW_JSON")]
    pub allow_json: bool,

    /// If present, `video/*` MIME types will be allowed
    #[clap(long = "allow-video", env = "CAMO_ALLOW_VIDEO")]
    pub allow_video: bool,

    /// The string used to identify this instance in upstream requests in Via and User-Agent
    #[clap(
        long = "header-via",
        env = "CAMO_HEADER_VIA",
        default_value = "camo-rs asset proxy (+https://github.com/denschub/camo-rs)"
    )]
    pub header_via: String,

    /// Randomly generated string used as a key for calculating the HMAC digest
    #[clap(long = "key", env = "CAMO_KEY")]
    pub key: String,

    /// The maximum `content-length`
    #[clap(
        long = "length-limit",
        env = "CAMO_LENGTH_LIMIT",
        default_value_t = 52428800
    )]
    pub length_limit: usize,

    /// IP and Port this instance should listen on
    #[clap(long = "listen", env = "CAMO_LISTEN", default_value = "[::]:8081")]
    pub listen: String,

    /// Defines how the log output will be formatted
    #[clap(value_enum, long = "log-format", env = "CAMO_LOG_FORMAT", default_value_t = LogFormat::TextColor)]
    pub log_format: LogFormat,

    /// The amount of log output. See the documentation for details
    #[clap(value_enum, long = "log-level", env = "CAMO_LOG_LEVEL", default_value_t = LogLevel::Quiet)]
    pub log_level: LogLevel,

    /// URL, including a trailing slash, relative to the domain Camo is running
    /// on
    ///
    /// For example, if Camo is available on `example.com/camo/`, set this
    /// to `/camo/`. For installations that do not run in a subdirectory, set
    /// this to `/`.
    #[clap(long = "root-url", env = "CAMO_ROOT_URL")]
    pub root_url: String,

    /// The number of seconds to wait for an upstream response
    #[clap(
        long = "upstream-timeout",
        env = "CAMO_UPSTREAM_TIMEOUT",
        default_value_t = 10
    )]
    pub upstream_timeout: usize,
}
